test/servinge2e/servicemesh_test.go

```diff
@@ -190,3 +231,295 @@ func TestKsvcWithServiceMeshSidecar(t *testing.T) {
+				"principalBinding": "USE_ORIGIN",
+				"origins": []map[string]interface{}{
+					{
+						"jwt": map[string]interface{}{
+							"issuer":  issuer,
+							"jwksUri": jwksKsvc.Status.URL,
+							"triggerRules": []map[string]interface{}{
+								{
+									"excludedPaths": []map[string]interface{}{
+										{
+											"prefix": "/metrics",
+										},
+										{
+											"prefix": "/healthz",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	policyGvr := schema.GroupVersionResource{
+		Group:    "authentication.istio.io",
+		Version:  "v1alpha1",
+		Resource: "policies",
+	}
+
+	authentication, err = caCtx.Clients.Dynamic.Resource(policyGvr).Namespace(serviceMeshTestNamespaceName).Create(authentication, meta.CreateOptions{})
+	if err != nil {
+		t.Fatalf("error creating istio Policy: %v", err)
+	}
+
+	caCtx.AddToCleanup(func() error {
+		t.Logf("Cleaning up istio Policy default")
+		return caCtx.Clients.Dynamic.Resource(policyGvr).Namespace(authentication.GetNamespace()).Delete(authentication.GetName(), &meta.DeleteOptions{})
+	})
+
+	// Create a test ksvc, should be accessible only via proper JWT token
+	testKsvc := test.Service("jwt-test", serviceMeshTestNamespaceName, image, map[string]string{
+		"sidecar.istio.io/inject": "true",
+	})
+	testKsvc = withServiceReadyOrFail(caCtx, testKsvc)
+
+	// Wait until the Route is ready and also verify the route returns a 401 without a token
+	if _, err := pkgTest.WaitForEndpointState(
+		&pkgTest.KubeClient{Kube: caCtx.Clients.Kube},
+		t.Logf,
+		testKsvc.Status.URL.URL(),
+		func(resp *spoof.Response) (bool, error) {
+			if resp.StatusCode != 401 {
+				// Returning (false, nil) causes SpoofingClient.Poll to retry.
+				return false, nil
+			}
+			return true, nil
+		},
+		"WaitForRouteToServe401",
+		true); err != nil {
+		t.Fatalf("the Route at domain %s didn't serve the expected HTTP 401 status: %v", testKsvc.Status.URL.URL(), err)
+	}
+
+	// Generate a valid JWT
+	validJwtToken, err := jwtRs256Token(privateKey, map[string]interface{}{
+		"iss": issuer,
+		"sub": subject,
+		"foo": "bar",
+		"iat": time.Now().Unix(),
+		"exp": time.Now().Unix() + 3600,
+	})
+	if err != nil {
+		t.Fatalf("error generating RS256 token: %v", err)
+	}
+
+	// Verify a request with a valid token receives HTTP 200 response with the expected content
+	resp, body, err := jwtHttpGetRequestBytes(testKsvc.Status.URL.String(), validJwtToken)
+	if err != nil {
+		t.Fatalf("error doing HTTP GET request: %v", err)
+	}
+
+	if resp.StatusCode != 200 || !strings.Contains(string(body), helloworldText) {
+		t.Fatalf("unexpected response with a valid token: HTTP %d: %s", resp.StatusCode, string(body))
+	}
+
+	// Now verify a few scenarios of invalid tokens
+
+	// Generate a new key for a "wrong key" scenario
+	wrongKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatalf("error generating private key: %v", err)
+	}
+
+	tests := []struct {
```

Comment:
I'd say the test above could be added to the table test below as well. The table can also define whether it expects success or failure so all tests could be there.

Reply:
That's how it originally was... however, the problem with that is if something is broken in such way that the token is not accepted at all,  all the negative tests will trivially Pass  (as they get the expected 401).

So then you see 3/4 tests passing, with just one failure (the test with the valid token)... 

So IMHO it is better to fail fully if the valid token is not accepted...

Reply:
In that case you could probably just call `return` or `break` in the given iteration and skip the other tests? Not sure whether we can check if the token was accepted at all. If yes, it would be good to add the check.

Reply:
not in the given iteration (t.Run runs in a separate goroutine) . We could conditionally not run the t.Run, but that would mean breaking the separation if the individual tests (leaking test result into the upper test logic), which I think would be an antipattern...

Reply:
Basically, in case the valid token is not accepted, testing whether any other _invalid_ token is accepted is pointless... that's why we skip the whole "testing of invalid tokens" section in that case...

Reply:
One possible alternative I can see would be to move the "valid token" test into each of the "invalid token" test scenario, instead of being a separate test... that would make all of the "invalid token" scenarios valid even on their own...

Reply:
IMO, it's not necessary to skip all negative tests if the "valid token" test fails. You see them dependent on each other but they're still valid individually. If the valid token test fails you still want the other negative tests to pass and make sure that invalid tokens are not accepted. At this point you don't care about the valid token but the invalid ones. I don't see those tests to be so tight together. This scenario is even more unlikely but: let's see the valid token test fails, and then the system accepts invalid token and returns success but we're skipping the test so we don't know that. I'd say it's better to keep all tests running. But don't insist one that much.

Reply:
OK, I have moved the valid token scenario to the list.